examples/seismic/skew_self_adjoint/utils.py

```diff
@@ -1,82 +1,159 @@
     """
     # sanity checks
     assert w > 0, "supplied w value [%f] must be positive" % (w)
     assert qmin > 0, "supplied qmin value [%f] must be positive" % (qmin)
     assert qmax > 0, "supplied qmax value [%f] must be positive" % (qmax)
     assert npad > 0, "supplied npad value [%f] must be positive" % (npad)
-    for n in wOverQ.data.shape:
+    for n in wOverQ.grid.shape:
         if n - 2*npad < 1:
             raise ValueError("2 * npad must not exceed dimension size!")
 
-    t1 = timer()
-    sigma = sigma or npad//11
     lqmin = np.log(qmin)
     lqmax = np.log(qmax)
 
-    if len(wOverQ.data.shape) == 2:
-        # 2d operations
-        nx, nz = wOverQ.data.shape
-        kxMin, kzMin = np.ogrid[:nx, :nz]
-        kxArr, kzArr = np.minimum(kxMin, nx-1-kxMin), np.minimum(kzMin, nz-1-kzMin)
-        nval1 = np.minimum(kxArr, kzArr)
-        nval2 = np.minimum(1, nval1/(npad))
-        nval3 = np.exp(lqmin+nval2*(lqmax-lqmin))
-        wOverQ.data[:, :] = w / nval3
+    # 1. Get distance to closest boundary in all dimensions
+    # 2. Logarithmic variation between qmin, qmax across the absorbing boundary
+    pos = Min(1, Min(*[Min(d - d.symbolic_min, d.symbolic_max - d)
+                       for d in wOverQ.dimensions]) / npad)
+    val = exp(lqmin + pos * (lqmax - lqmin))
+
+    # 2020.05.04 currently does not support spatial smoothing of the Q field
+    # due to MPI weirdness in reassignment of the numpy array
+    eqn1 = Eq(wOverQ, w / val)
+    Operator([eqn1], name='WOverQ_Operator')()
+
+#     eqn1 = Eq(wOverQ, val)
+#     Operator([eqn1], name='WOverQ_Operator_init')()
+#     # If we apply the smoother, we must renormalize output to [qmin,qmax]
+#     if sigma > 0:
+#         print("sigma=", sigma)
+#         smooth = gaussian_smooth(wOverQ.data, sigma=sigma)
+#         smin, smax = np.min(smooth), np.max(smooth)
+#         smooth[:] = qmin + (qmax - qmin) * (smooth - smin) / (smax - smin)
+#         wOverQ.data[:] = smooth
+#     eqn2 = Eq(wOverQ, w / wOverQ)
+#     Operator([eqn2], name='WOverQ_Operator_recip')()
 
+
+def defaultSetupIso(npad, shape, dtype,
+                    sigma=0, qmin=0.1, qmax=100.0, tmin=0.0, tmax=2000.0,
+                    bvalue=1.0/1000.0, vvalue=1.5, space_order=8):
+    """
+    For isotropic propagator build default model with 10m spacing,
+        and 1.5 m/msec velocity
+
+    Return:
+        dictionary of velocity, buoyancy, and wOverQ
+        TimeAxis defining temporal sampling
+        Source locations: one located at center of model, z = 1dz
+        Receiver locations, one per interior grid intersection, z = 2dz
+            2D: 1D grid of receivers center z covering interior of model
+            3D: 2D grid of receivers center z covering interior of model
+    """
+    d = 10.0
+    origin = tuple([0.0 - d * npad for s in shape])
+    extent = tuple([d * (s - 1) for s in shape])
+
+    # Define dimensions
+    if len(shape) == 2:
```

Comment:
nitpicking: I think with a list comprehension computing `x,y,z` (or just `x,y`) we can get rid of this `if`

Reply:
you may have to show me what you mean im a python n00b

Reply:
```
dimensions = ['x', 'z', 'y'][:len(shape)]
dimensions = [SpaceDimension(name=i, spacing=Constant(name='h_%s' % i, value=d)) for i in dimensions]
grid = Grid(..., dimensions=dimensions, ...)
```

Reply:
did you also mean the second conditional below, for instrument locations? I suppose all of this might be moot when I refactor to use `Model` and `AcquistiionGeometry`

Reply:
Will defer, as use of model/geom will remove this issue.